pkg/controller/cloud/nodecontroller.go

```diff
@@ -155,3 +267,121 @@ func (cnc *CloudNodeController) Run() {
 		}, cnc.nodeMonitorPeriod, wait.NeverStop)
 	}()
 }
+
+func (cnc *CloudNodeController) AddCloudNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	instances, ok := cnc.cloud.Instances()
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("cloudprovider does not support instances"))
+		return
+	}
+
+	// This initializes nodes with cloud info
+	// Only initializes nodes that were created with the "ExternalCloudProvider" taint
+	taints := node.Spec.Taints
+
+	var cloudTaint *v1.Taint
+	for _, taint := range taints {
+		glog.Errorf("Processing taint %s, comparing with %s", taint.Key, CloudTaintKey)
+		if taint.Key == CloudTaintKey {
+			cloudTaint = &taint
+		}
+	}
+
+	if cloudTaint == nil {
+		glog.V(2).Infof("This node %s is registered without the cloud taint. Will not process.", node.Name)
+		return
+	}
+
+	err := clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
+		curNode, err := cnc.kubeClient.Core().Nodes().Get(node.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// If user provided an IP address, ensure that IP address is found
+		// in the cloud provider before removing the taint on the node
+		var nodeIP net.IP
+		if ip, ok := node.ObjectMeta.Labels[LabelProvidedIPAddr]; ok {
+			nodeIP = net.ParseIP(ip)
+		}
+		if nodeIP != nil {
+			nodeAddresses, err := instances.NodeAddressesByProviderID(node.Spec.ProviderID)
+			if err != nil {
+				nodeAddresses, err = instances.NodeAddresses(types.NodeName(node.Name))
+				if err != nil {
+					glog.Errorf("failed to get node address from cloud provider: %v", err)
+					return nil
+				}
+			}
+			var providedIP *v1.NodeAddress
+			for i := range nodeAddresses {
+				if nodeAddresses[i].Address == nodeIP.String() {
+					providedIP = &nodeAddresses[i]
+				}
+			}
+			if providedIP == nil {
+				glog.Errorf("failed to get node address for node %s from cloudprovider that matches ip: %v", node.Name, nodeIP)
+				return nil
+			}
+		}
+
+		instanceType, err := instances.InstanceTypeByProviderID(curNode.Spec.ProviderID)
+		if err != nil {
+			instanceType, err = instances.InstanceType(types.NodeName(curNode.Name))
+			if err != nil {
+				glog.Errorf("Error getting instance type %v", err)
+				return err
+			}
+		}
+		if instanceType != "" {
+			glog.Infof("Adding node label from cloud provider: %s=%s", metav1.LabelInstanceType, instanceType)
+			curNode.ObjectMeta.Labels[metav1.LabelInstanceType] = instanceType
+		}
+
+		// Since there are node taints, do we still need this?
+		// This condition marks the node as unusable until routes are initialized in the cloud provider
+		if cnc.cloud.ProviderName() == "gce" {
```

Comment:
We do need this. The routecontroller looks for this condition to identify nodes that need a route.